File: packages/android/src/androidTest/kotlin/dev/sinua/core/FxSpecTests.kt
package dev.sinua.core

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import uniffi.core_engine.checkOverrides
import uniffi.core_engine.estimateCost
import uniffi.core_engine.frameFromFxSpec
import uniffi.core_engine.frameFromFxSpecWith
import uniffi.core_engine.frameWithOverrides
import uniffi.core_engine.fxColorToHsl
import uniffi.core_engine.fxSpecCost
import uniffi.core_engine.parameterCatalogJson
import uniffi.core_engine.particleDefaults
import uniffi.core_engine.resolveFxSpec
import uniffi.core_engine.resolveFxSpecWith
import uniffi.core_engine.resolvedOpts

/**
 * FX Spec through the UniFFI boundary on an Android runtime: every
 * `spec/examples/` spec resolves with no errors, and `frameFromFxSpec`
 * renders exactly what `frameWithOverrides(resolved)` at
 * `elapsed * presetSpeed * spec.speed` does -- the property
 * `crates/core_engine/src/fx_spec.rs` checks natively. The examples ride in
 * the test APK via the same `androidTest` assets dir as the golden file
 * (`spec/`, so they sit under `examples/`).
 */
@RunWith(AndroidJUnit4::class)
class FxSpecTests {
    @Test
    fun examplesRenderLikeOverrides() {
        val assets = InstrumentationRegistry.getInstrumentation().context.assets
        val files = assets.list("examples")!!.filter { it.endsWith(".fxspec.json") }.sorted()
        assertTrue("found ${files.size} examples", files.size >= 8)
        for (file in files) {
            val json = assets.open("examples/$file").bufferedReader().use { it.readText() }
            // v1.1: every `states` key too, with the fixed inputs every
            // platform's parity test passes (fx-spec.test.mjs's TEST_INPUTS).
            val keys: List<String?> = listOf(null) + resolveFxSpec(json).stateKeys
            for (key in keys) {
                for (low in listOf(false, true)) {
                    val label = "$file ${key ?: "<base>"} lowPower=$low"
                    val r = resolveFxSpecWith(json, key, INPUTS, low)
                    assertTrue("$label: ${r.diagnostics}", r.ok)
                    assertEquals(label, emptyList<String>(), r.inactiveBindings)
                    val preset = resolvedOpts(r.state, r.size)!!.speed
                    val got = frameFromFxSpecWith(json, 1.3, key, INPUTS, low)
                    assertNotNull(label, got)
                    assertEquals(label, frameWithOverrides(r.state, r.size, 1.3 * preset * r.speed, r.overrides), got)
                    assertNotNull(label, fxSpecCost(json, key, INPUTS, low))
                }
            }
        }
        // FX Spec 1.2: low power caps fps and sheds glow + noise.
        val power = assets.open("examples/status-beacon-power.fxspec.json").bufferedReader().use { it.readText() }
        assertEquals(30.0, resolveFxSpecWith(power, null, emptyMap()).maxFps)
        val low = resolveFxSpecWith(power, null, emptyMap(), true)
        assertEquals(15.0, low.maxFps)
        assertEquals(listOf("glow", "noise"), low.disabledMaterials)
        assertEquals("heavy", estimateCost("working", 64u, mapOf("glowStrength" to 1.0))?.`class`)
        // Per-state particle defaults (the Studio's knob defaults).
        assertEquals(3.0, particleDefaults("tracking")?.get("particleStyle"))
        assertEquals(1.0, particleDefaults("notifying")?.get("particleSync"))
        assertEquals(4.5, particleDefaults("working")?.get("particleLife"))
        assertNull(particleDefaults("nope"))
    }

    private companion object {
        val INPUTS = mapOf(
            "micMuted" to 0.0,
            "micLevel" to 0.6,
            "agentVolume" to 0.5,
            "steps" to 6200.0,
            "waterMl" to 1800.0,
            "activeMinutes" to 12.0,
            "heartRate" to 128.0,
        )
    }

    @Test
    fun diagnosticsAndColor() {
        val r = resolveFxSpec("""{"fxSpec":"1.8","object":"orb","pattern":"working","params":{"orbitn":3}}""")
        assertFalse(r.ok)
        assertTrue(r.diagnostics.any { it.path == "/params/orbitn" && it.message.contains("`orbitN`") })
        assertNull(frameFromFxSpec("""{"fxSpec":"2.0","object":"orb","pattern":"working"}""", 0.0))
        val c = fxColorToHsl("#ff00ff")!!
        assertEquals("#ff00ff", c.hex)
        assertEquals(300.0, c.h, 1e-9)
    }

    /** The parameter catalog (docs/parameters.md) crosses UniFFI as JSON text; checkOverrides warns only. */
    @Test
    fun parameterCatalogAndCheckOverrides() {
        val cat = org.json.JSONObject(parameterCatalogJson())
        val objects = cat.getJSONArray("objects")
        assertEquals(
            listOf("SinuaOrb", "SinuaSignal", "SinuaRing", "SinuaCore", "SinuaBeacon"),
            (0 until objects.length()).map { objects.getJSONObject(it).getString("component") },
        )
        assertEquals(
            "glow.strength",
            cat.getJSONObject("definitions").getJSONObject("glowStrength@shared").getString("path"),
        )
        val w = checkOverrides("breathing", 64u, mapOf("lanse" to 6.0))
        assertEquals(1, w.size)
        assertEquals("warning", w[0].severity)
        assertTrue(w[0].message.contains("did you mean `lanes`"))
        assertTrue(checkOverrides("tracking", 64u, mapOf("progress1" to 2.0)).isEmpty())
    }
}

File: packages/ios/Tests/CoreEngineTests/FxSpecTests.swift
import Foundation
import XCTest

@testable import CoreEngine

/// FX Spec through the UniFFI boundary on Apple's runtime: every
/// `spec/examples/*.fxspec.json` resolves with no errors, and
/// `frameFromFxSpec` renders exactly what `frameWithOverrides(resolved)` at
/// `elapsed * presetSpeed * spec.speed` does -- the property
/// `crates/core_engine/src/fx_spec.rs` checks natively and
/// `packages/core/test/fx-spec.test.mjs` checks on wasm. Read via
/// `#filePath` like `SinuaGoldenTests`, skipping where the repo isn't
/// visible (device runs).
final class FxSpecTests: XCTestCase {
    private func examplesDir() -> URL {
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()  // CoreEngineTests
            .deletingLastPathComponent()  // Tests
            .deletingLastPathComponent()  // ios
            .deletingLastPathComponent()  // packages
            .deletingLastPathComponent()  // repo root
            .appendingPathComponent("spec/examples")
    }

    func testExamplesRenderLikeOverrides() throws {
        let dir = examplesDir()
        guard let files = try? FileManager.default.contentsOfDirectory(atPath: dir.path) else {
            throw XCTSkip("spec/examples not readable at \(dir.path) (device run?)")
        }
        let specs = files.filter { $0.hasSuffix(".fxspec.json") }.sorted()
        XCTAssertGreaterThanOrEqual(specs.count, 8)
        for file in specs {
            let json = try String(contentsOf: dir.appendingPathComponent(file), encoding: .utf8)
            // v1.1: every `states` key too, with the fixed inputs every
            // platform's parity test passes (fx-spec.test.mjs's TEST_INPUTS).
            let keys: [String?] = [nil] + CoreEngine.resolveFxSpec(json: json).stateKeys
            for key in keys {
                for low in [false, true] {
                    let label = "\(file) \(key ?? "<base>") lowPower=\(low)"
                    let r = CoreEngine.resolveFxSpecWith(json: json, state: key, inputs: Self.inputs, lowPower: low)
                    XCTAssertTrue(r.ok, "\(label): \(r.diagnostics)")
                    XCTAssertEqual(r.inactiveBindings, [], label)
                    let preset = try XCTUnwrap(CoreEngine.resolvedOpts(state: r.state, size: r.size)).speed
                    let got = CoreEngine.frameFromFxSpecWith(
                        json: json, elapsed: 1.3, state: key, inputs: Self.inputs, lowPower: low)
                    let want = CoreEngine.frameWithOverrides(
                        state: r.state, size: r.size, t: 1.3 * preset * r.speed, overrides: r.overrides)
                    XCTAssertNotNil(got, label)
                    XCTAssertEqual(got, want, label)
                    // The cost estimate is available natively too.
                    XCTAssertNotNil(
                        CoreEngine.fxSpecCost(json: json, state: key, inputs: Self.inputs, lowPower: low), label)
                }
            }
        }
        // FX Spec 1.2: low power caps fps and sheds glow + noise.
        let power = try String(
            contentsOf: dir.appendingPathComponent("status-beacon-power.fxspec.json"), encoding: .utf8)
        XCTAssertEqual(CoreEngine.resolveFxSpecWith(json: power, state: nil, inputs: [:]).maxFps, 30)
        let low = CoreEngine.resolveFxSpecWith(json: power, state: nil, inputs: [:], lowPower: true)
        XCTAssertEqual(low.maxFps, 15)
        XCTAssertEqual(low.disabledMaterials, ["glow", "noise"])
        XCTAssertEqual(
            CoreEngine.estimateCost(state: "working", size: 64, overrides: ["glowStrength": 1])?.class, "heavy")
        // Per-state particle defaults (the Studio's knob defaults).
        XCTAssertEqual(CoreEngine.particleDefaults(state: "tracking")?["particleStyle"], 3)
        XCTAssertEqual(CoreEngine.particleDefaults(state: "notifying")?["particleSync"], 1)
        XCTAssertEqual(CoreEngine.particleDefaults(state: "working")?["particleLife"], 4.5)
        XCTAssertNil(CoreEngine.particleDefaults(state: "nope"))
    }

    private static let inputs: [String: Double] = [
        "micMuted": 0, "micLevel": 0.6, "agentVolume": 0.5, "steps": 6200, "waterMl": 1800,
        "activeMinutes": 12, "heartRate": 128,
    ]

    func testDiagnosticsAndColor() {
        let r = CoreEngine.resolveFxSpec(
            json: #"{"fxSpec":"1.8","object":"orb","pattern":"working","params":{"orbitn":3}}"#)
        XCTAssertFalse(r.ok)
        XCTAssertTrue(r.diagnostics.contains { $0.path == "/params/orbitn" && $0.message.contains("`orbitN`") })
        XCTAssertNil(
            CoreEngine.frameFromFxSpec(json: #"{"fxSpec":"2.0","object":"orb","pattern":"working"}"#, elapsed: 0))
        let c = CoreEngine.fxColorToHsl(color: "#ff00ff")
        XCTAssertEqual(c?.hex, "#ff00ff")
        XCTAssertEqual(c?.h ?? -1, 300, accuracy: 1e-9)
    }

    /// The parameter catalog (docs/parameters.md) crosses UniFFI as JSON text;
    /// checkOverrides returns warnings only.
    func testParameterCatalogAndCheckOverrides() throws {
        let data = CoreEngine.parameterCatalogJson().data(using: .utf8)!
        let cat = try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any])
        let objects = try XCTUnwrap(cat["objects"] as? [[String: Any]])
        XCTAssertEqual(
            objects.compactMap { $0["component"] as? String },
            ["SinuaOrb", "SinuaSignal", "SinuaRing", "SinuaCore", "SinuaBeacon"])
        let defs = try XCTUnwrap(cat["definitions"] as? [String: [String: Any]])
        XCTAssertEqual(defs["glowStrength@shared"]?["path"] as? String, "glow.strength")
        let w = CoreEngine.checkOverrides(state: "breathing", size: 64, overrides: ["lanse": 6])
        XCTAssertEqual(w.count, 1)
        XCTAssertEqual(w.first?.severity, "warning")
        XCTAssertTrue(w.first?.message.contains("did you mean `lanes`") ?? false)
        XCTAssertTrue(CoreEngine.checkOverrides(state: "tracking", size: 64, overrides: ["progress1": 2]).isEmpty)
    }
}
